### PR TITLE
Feature/checksum

### PR DESCRIPTION
**Pull request by LattaruoloAndrea**
Added the features for checksum and the incomplete file mentioned in [#69](https://github.com/SDFIdk/CDSETool/issues/69)  [#40](https://github.com/SDFIdk/CDSETool/issues/40)

### CODE REVIEW

**Comment by LattaruoloAndrea** [comment]
> @LattaruoloAndrea Thank you very much for your contribution 🥇
> 
> I can see quite a bit of code commented out, and other bits left `# TODO`. Would you like to clean up the code a bit, or would you prefer i give it a go?

@pedep  I am unsure on some of the parts (especially the ones in which there are different fields to handle), maybe I can try to clean the code that I implemented and then make a new pull request and you can have a look to that one.

**Comment by pedep** [comment]
@LattaruoloAndrea Sounds good :)

Just let me know if you need help implementing, or wish to discuss along the way :+1:

File: src/cdsetool/download.py [hunk 1]
```diff
@@ -28,37 +36,100 @@ def download_feature(feature, path, options=None):
...
+                    if algo == "sha3-256":
+                        checksum = checksum_dict['Value']
+                        algo = hashlib.sha3_256() 
+                    elif algo == "MD5":
+                        checksum = checksum_dict['Value']
+                        algo = hashlib.md5()
+                        break
+                    elif algo == "BLAKE3":
+                        checksum = checksum_dict['Value']
+                        algo = hashlib.blake2b() # TODO ?
```
Comment (pedep) [inline src/cdsetool/download.py:113]: There is no reason to loop over each algorithm. We just need the first
```python3
checksums = product_info.get("Checksum", [])
if len(checksums) == 0:
  return True # no checksum available, so we skip verifying

algo, checksum = checksums[0].get("Algorithm"), checksums[0].get("Value")

algo = {
    "sha3-256": hashlib.sha3_256(),
    "MD5": hashlib.md5(),
    "BLAKE3": blake3() # https://pypi.org/project/blake3/
}[algo]

File: src/cdsetool/download.py [hunk 2]
```diff
@@ -28,37 +36,100 @@ def download_feature(feature, path, options=None):
...
+    #     status.set_filename(f"{algo} - {filename}")
+    #     status.set_filesize(file_size)
+    with open(file_path, "rb") as f:
+        while True:
+            block_data = f.read(block_size)
+            if not block_data:
+                break
+            algo.update(block_data)
+            # status.add_progress(block_data)
+    return algo.hexdigest().lower() == checksum.lower()
```
Comment (pedep) [inline src/cdsetool/download.py:132]: If you move this part into the download loop, you can avoid having to read the file again. You can essentially halve the amount of disk IO for each file.

Heres a simplified example
```python3
        checksum = hashlib.md5() # select the correct algo
        with open(temp_path, "wb") as file:
            for chunk in response.iter_content(chunk_size=1024 * 1024 * 5):
                if chunk:
                    file.write(chunk)
                    checksum.update(chunk)
                    status.add_progress(len(chunk))
        assert checksum.digest() == the_odata_checksum

File: src/cdsetool/download.py [hunk 3]
```diff
@@ -105,10 +174,20 @@ def _retry_backoff(url, session):
 
     return response
 
+def _set_proxy(options,session):
+    proxies = options.get("proxies", {})
+    if proxies != {}:
+        if "http" in proxies or "https" in proxies:
```
Comment (pedep) [inline src/cdsetool/download.py:180]: This line does nothing. `requests` has the same check internally, so theres no reason for us to repeat it
```suggestion
```

**Review by pedep — CHANGES_REQUESTED** [review]
Good work  :+1: 
I have left a few comments with some wishes for refactoring.

For the future, i would much prefer if you kept features in separate pull requests. This makes it easier for me to review the code for mistakes and quality :nerd_face:

**Comment by LattaruoloAndrea** [comment]
Thanks for the comments I will try to work on it, next time I will try to work on just one feature at the time.

File: src/cdsetool/query.py [hunk 4]
```diff
@@ -63,12 +63,73 @@ def __fetch_features(self):
...
+                    plevel = original_filters.get("processingLevel",None)
+                    odata_query = self.format_query(start_date= start_date,end_date= end_date,data_collection= data_collection,aoi= aoi,order_by= order_by,product_type=product_type,mode=mode,direction=direction,orbit=orbit,tileid=tileid,plevel=plevel)
+                    if odata_query.strip() == "":
+                        # An empty query should return the full set of products on the server, which is a bit unreasonable.
+                        # The server actually raises an error instead and it's better to fail early in the client.
+                        raise ValueError("Empty query.")
+                    res = self.__query_call_ordered_dict(url=odata_url,query=odata_query)
+                    return res
+        #one of the filters is missing
+        return {}
```
[the remark this answers is not in the document]
Comment (LattaruoloAndrea) [reply src/cdsetool/query.py]: This sounds great, to fetch the data by listing the procuts title should work now it's simplified like this
```Python
    def __fetch_odata_from_product_list(self,product_list):
        """
        Given a product list, return the odata response
        """
        body = {}
        list_products = []
        for product in product_list:
            if 'properties' in product:
                if 'title' in product['properties']:
                    list_products.append({"Name": product['properties']['title']})
        body['FilterProducts'] = list_products
        odata_url = "https://catalogue.dataspace.copernicus.eu/odata/v1/Products/OData.CSC.FilterList"
        res = self.__get_odata(odata_url,body)
        return res
    
    def __get_odata(self,url,data_json):
        """
        From the odata url and request body, get the odata response
        """
        res = requests.post(url,json=data_json, timeout=120).json()
        return res
```

File: src/cdsetool/query.py [hunk 5]
```diff
@@ -63,12 +63,73 @@ def __fetch_features(self):
...
             res = requests.get(self.next_url, timeout=120).json()
             self.features += res.get("features") or []
-
             total_results = res.get("properties", {}).get("totalResults")
             if total_results is not None:
                 self.total_results = total_results
-
+            odata =self.__fetch_odata(res.get("properties", {}))
+            if odata != {}:
+                self.__add_odata_to_features(odata)
```
[the remark this answers is not in the document]
Comment (LattaruoloAndrea) [reply src/cdsetool/query.py]: Here I have made a few changes, I like the idea of querying the odata in the query function. I have implemented two versions, if in the query_features function parameters we pass the True at the end we get the query with oData inside.
```Python
query_features(
    "Sentinel1",
    {
        "startDate": "2023-12-24",
        "completionDate": date(2024, 1, 4),
        "processingLevel": "LEVEL1",
        "sensorMode": "IW",
        "productType": "IW_GRDH_1S",
        # "orbitDirection":"dir",
        # "orbitNumber":13,
    },
    True #here the new parameter
)
```
Then since it might be useful for some people to check the odata I saved the oData inside a field in featues like this:
```Python
    def __add_odata_to_features(self,odata):
        """
        Given the odatadict, add the to correct (by matching the Ids) checksum the features
        """
        map_id_checksum = {}
        if "value" in odata:
            odata_list_value = odata["value"]
            for ordered_dict in odata_list_value:
                map_id_checksum[ordered_dict["Id"]] = {"Checksum": ordered_dict["Checksum"], "ContentLength": ordered_dict["ContentLength"],"odata": ordered_dict}
        for feature in self.features:
            if feature.get("id") in map_id_checksum.keys():
                ordered_dict = map_id_checksum[feature.get("id")]
                feature['odata'] = ordered_dict["odata"]
                feature['Checksum'] = ordered_dict["Checksum"]
                if feature.get('properties').get('services').get('download').get('size') == ordered_dict['ContentLength']:
                    feature['ContentLength'] = ordered_dict['ContentLength']
                else:
                    print(f"Warning: content length of {feature.get('id')} does not match the ContentLength in odata")
            else:
                print(f"Warning: {feature.get('id')} not in odata")
```
In which we still add the checksum and the oData (inside oData it will still be present the checksum). Let me know if you like this implementation or not. At the same time if you don't specify the "True" inside the query_featues I will fetch the oData during the checksum validate funtion like this:
```Python
def _checksum_compare(temp_path, product_info):
    """Compare a given MD5 checksum with one calculated from a file."""
    checksum = None
    algo = None
    filename = product_info.get("properties").get("title")
    if "Checksum" in product_info:
        # The intial query was made with the checksum option.
        checksum_list =product_info.get("Checksum")
        algo, checksum =from_checksum_list_to_checksum(checksum_list)
    else:
        odata_response = get_odata_from_product_title(filename)
        if 'value' in odata_response:
            if len(odata_response['value']) ==1:
                # odata['value'] is a list if there is more then one element for a single product
                # there is a error on the odata I guess 
                if 'Checksum' in odata_response['value'][0]:
                    checksum_list = odata_response['value'][0]['Checksum']
                    algo, checksum =from_checksum_list_to_checksum(checksum_list)
```
basically it checks wheter checksum is present in the options otherwise fetch the odata

File: src/cdsetool/download.py [hunk 6]
```diff
@@ -28,37 +36,100 @@ def download_feature(feature, path, options=None):
         return feature.get("id")
 
     overwrite = options.get("overwrite_existing", False)
+    continue_ = options.get("continue_existing", True)
 
+    if not os.path.exists(Path(path)):
+        os.makedirs(path)
     result_path = os.path.join(path, filename.replace(".SAFE", ".zip"))
-
+    temp_path =  Path(path) / filename.replace(".SAFE", ".zip.incomplete")
```
[the remark this answers is not in the document]
Comment (LattaruoloAndrea) [reply src/cdsetool/download.py]: I changed the Path() to os.path.join, since I am used to use that but did't see any good reason for not to use o.path.join()

File: src/cdsetool/download.py [hunk 7]
```diff
@@ -28,37 +36,100 @@ def download_feature(feature, path, options=None):
         return feature.get("id")
 
     overwrite = options.get("overwrite_existing", False)
+    continue_ = options.get("continue_existing", True)
```
[the remark this answers is not in the document]
Comment (LattaruoloAndrea) [reply src/cdsetool/download.py]: yeah sounds great

File: src/cdsetool/download.py [hunk 8]
```diff
@@ -28,37 +36,100 @@ def download_feature(feature, path, options=None):
...
+                        checksum = checksum_dict['Value']
+                        algo = hashlib.md5()
+                        break
+                    elif algo == "BLAKE3":
+                        checksum = checksum_dict['Value']
+                        algo = hashlib.blake2b() # TODO ?
+                else:
+                    raise InvalidChecksumError("No checksum information found in product information. No Algorithm provided provided")
+        else:
+                    raise InvalidChecksumError("No checksum information found in product information. The Checksum list is empty")
```
[the remark this answers is not in the document]
Comment (LattaruoloAndrea) [reply src/cdsetool/download.py]: I removed all the raise that I did introduce for the checksum. so now it shold fail silently

File: src/cdsetool/download.py (hunk 1 above)
Comment (LattaruoloAndrea) [reply src/cdsetool/download.py]: I really like your implementations it's clean here it's the change for the _checksum_compare()
```Python
def _checksum_compare(temp_path, product_info):
    """Compare a given MD5 checksum with one calculated from a file."""
    checksum = None
    algo = None
    filename = product_info.get("properties").get("title")
    if "Checksum" in product_info:
        # The intial query was made with the checksum option.
        checksum_list =product_info.get("Checksum")
        algo, checksum =from_checksum_list_to_checksum(checksum_list)
    else:
        odata_response = get_odata_from_product_title(filename)
        if 'value' in odata_response:
            if len(odata_response['value']) ==1:
                # odata['value'] is a list if there is more then one element for a single product
                # there is a error on the odata I guess 
                if 'Checksum' in odata_response['value'][0]:
                    checksum_list = odata_response['value'][0]['Checksum']
                    algo, checksum =from_checksum_list_to_checksum(checksum_list)
    if checksum == None:
        return None # no checksum available, so we skip verifying
    file_path = temp_path
    with open(file_path, "rb") as f:
        while True:
            block_data = f.read(8192)
            if not block_data:
                break
            algo.update(block_data)
            # status.add_progress(block_data)
    return algo.hexdigest().lower() == checksum.lower() # checksum available, so we give the verify checksum

def from_checksum_list_to_checksum(checksum_list):
    if len(checksum_list) == 0:
        return None, None # no checksum available, so we skip verifying
    algo, checksum = checksum_list[0].get("Algorithm"), checksum_list[0].get("Value")
    algo = {
        "sha3-256": hashlib.sha3_256(),
        "MD5": hashlib.md5(),
        "BLAKE3": blake3() # https://pypi.org/project/blake3/
    }[algo]
    return algo,checksum
```
Also I have added inside the pyproject.toml the following part, in which I add blake for the blake3 I don't know if it should go also somewhere else
```
dependencies = [
  "typer ~= 0.9.0",
  "rich >= 13.6,< 13.8",
  "requests >= 2.28.1,< 2.32.0",
  "pyjwt[crypto] ~= 2.8.0",
  "blake >=0.4.1" #here new depencency
]
```

File: src/cdsetool/download.py (hunk 2 above)
Comment (LattaruoloAndrea) [reply src/cdsetool/download.py]: Here I am insecure would this work for partial file as well? (we shoul add the part that was already downloaded before to the checksum in case). In this case this part I did't implenet for this reason.

File: src/cdsetool/download.py (hunk 3 above)
Comment (LattaruoloAndrea) [reply src/cdsetool/download.py]: Removed.

File: src/cdsetool/download.py [hunk 9]
```diff
@@ -7,19 +7,27 @@
...
+from src.cdsetool._processing import _concurrent_process
+from src.cdsetool.credentials import Credentials
+from src.cdsetool.monitor import NoopMonitor
 
+class InvalidChecksumError(Exception):
+    """MD5 checksum of a local file does not match the one from the server."""
+    pass
 
-def download_feature(feature, path, options=None):
+def download_feature(feature, path, options=None,try_number = 0):
```
[the remark this answers is not in the document]
Comment (LattaruoloAndrea) [reply src/cdsetool/download.py:24]: Done.

File: src/cdsetool/download.py [hunk 10]
```diff
@@ -7,19 +7,27 @@
...
     """
     Download a single feature
 
     Returns the feature ID
     """
+    if try_number ==10:
+        """Tries 10 times to download a partial file """
+        return feature.get("id")
```
[the remark this answers is not in the document]
Comment (LattaruoloAndrea) [reply src/cdsetool/download.py:32]: Done. Better this way

**Comment by LattaruoloAndrea** [comment]
Hello @pedep do you have any news on the changes? Do you want me to midofy something or create different pull requests for each feture that I implemented?

**Comment by pjonsson** [comment]
I'm not speaking for the project, but personally I have a hard time to look at the diff and determine why a specific line is changed.

I'm guessing the proxy change is small (5-10 lines?) and supporting a proxy shouldn't be controversial, so perhaps you can lift that change out in a separate PR against latest main and see if that's easier to review? Parts of this PR has what to me looks like whitespace damage, so make sure to run black on the source code before making the PR.
